src/live-programming.rst

```diff
@@ -20,9 +20,331 @@ Data Analysis Environments
+  completing their tasks in significantly less time than with an existing IDE.
+
+
+Algorithmic Debugging
+^^^^^^^^^^^^^^^^^^^^^
+
+*Algorithmic debugging* (also called *declarative debugging*) is a semi-automatic
+debugging technique in which the debugger automatically generates a series of
+questions to which the programmer's answers guide the search toward the bug.
+The debugger constructs an *execution tree* (ET), a data structure representing a
+program execution, and traverses it using some search strategy, asking the
+programmer at each ET node whether the represented subcomputation is correct
+in order to determine the next step.
+This technique guarantees that, if the programmer answers all the questions correctly,
+the bug will eventually be found.
+
+Although algorithmic debugging can be applied in any language paradigm, it is most
+suited for declarative languages, e.g., pure functional languages.
+To determine whether an ET node for a pure functional program is correct, the
+programmer need only check that the return value of the corresponding, referentially
+transparent expression is the expected one, independent of any other ET node.
+On the other hand, checking the correctness of an ET node for an imperative program
+requires checking, in addition to the return value, that values in the heap have
+been updated correctly---this can be difficult to ascertain because the programmer
+must maintain an understanding of how any *subsequent ET node* depends on those updated
+values.
+
+Despite their useful guarantee of bug diagnosis, algorithmic debuggers have yet to enter
+widespread use. (add note about functional programming research lagging in tooling)
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'nilsson1994lazy'
+
+  .. todo::
+    write bib entry
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'silva2006adps'
+
+  This paper combined program slicing and algorithmic debugging
+  into a unified debugging framework for functional programs, adapting
+  similar combinations applied to logic and imperative programs.
+  The paper was motivated by the authors' observation that AD tools would produce
+  long series of semantically loosely connected and, from the user's perspective,
+  redundant questions.
+  Program slicing provides a complementary remedy: rather than just 'correct'
+  or 'incorrect',
+  the user may provide a slicing criterion specifying which parts of the
+  subcomputation's result are incorrect; the slicing criterion is used to prune
+  the debugging tree of irrelevant subcomputations, leading to more semantically
+  connected questions.
+  The paper adapted program slicing concepts to the Augmented Redex Trail (ART), a
+  trace structure that can be the basis for both AD and program slicing, and presented
+  an algorithm for slicing ARTs.
+
+  .. note::
+    The debugging interface sketched in :cite:`perera2012explain` may be viewed
```

Comment:
maybe make a forward reference to algorithmic debugging in the summary for roly's paper above